osquery/sql/sqlite_util.cpp

```diff
@@ -459,41 +459,115 @@ Status QueryPlanner::applyTypes(TableColumns& columns) {
-      }
-      r[column[i]] = (argv[i] != nullptr) ? argv[i] : FLAGS_nullvalue;
+Status readRows(sqlite3_stmt* prepared_statement,
+                QueryDataTyped& results,
+                const SQLiteDBInstanceRef& instance) {
+  int rc = sqlite3_step(prepared_statement);
+  /* if we have a result set row... */
+  if (SQLITE_ROW == rc) {
+    // First collect the column names
+    int num_columns = sqlite3_column_count(prepared_statement);
+    std::vector<std::string> colNames;
+    colNames.reserve(num_columns);
+    for (int i = 0; i < num_columns; i++) {
+      colNames.push_back(sqlite3_column_name(prepared_statement, i));
     }
+
+    do {
+      RowTyped row;
+      for (int i = 0; i < num_columns; i++) {
+        switch (sqlite3_column_type(prepared_statement, i)) {
+        case SQLITE_INTEGER:
+          row[colNames[i]] =
+              static_cast<int64_t>(sqlite3_column_int64(prepared_statement, i));
+          break;
+        case SQLITE_FLOAT:
+          row[colNames[i]] = sqlite3_column_double(prepared_statement, i);
+          break;
+        case SQLITE_NULL:
+          row[colNames[i]] = FLAGS_nullvalue;
+          break;
+        default:
+          // Everything else (SQLITE_TEXT, SQLITE3_TEXT, SQLITE_BLOB) is
+          // obtained/conveyed as text/string
+          row[colNames[i]] = std::string(reinterpret_cast<const char*>(
+              sqlite3_column_text(prepared_statement, i)));
+        }
+      }
+      results.push_back(std::move(row));
+      rc = sqlite3_step(prepared_statement);
+    } while (SQLITE_ROW == rc);
+  }
+  if (!(rc == SQLITE_OK || rc == SQLITE_DONE)) {
+    return Status::failure(sqlite3_errmsg(instance->db()));
+  }
+
+  rc = sqlite3_finalize(prepared_statement);
+  if (rc != SQLITE_OK) {
+    return Status::failure(sqlite3_errmsg(instance->db()));
   }
-  (*qData).push_back(std::move(r));
-  return 0;
+
+  return Status::success();
 }
 
-Status queryInternal(const std::string& q,
-                     QueryData& results,
+Status queryInternal(const std::string& query,
+                     QueryDataTyped& results,
                      const SQLiteDBInstanceRef& instance) {
-  char* err = nullptr;
-  auto lock = instance->attachLock();
-  sqlite3_exec(instance->db(), q.c_str(), queryDataCallback, &results, &err);
+  sqlite3_stmt* prepared_statement{nullptr}; /* Statement to execute. */
+
+  int rc = SQLITE_OK; /* Return Code */
+  const char* leftover_sql = nullptr; /* Tail of unprocessed SQL */
+  const char* sql = query.c_str(); /* SQL to be processed */
+
+  /* The big while loop.  One iteration per statement */
+  while ((sql[0] != '\0') && (SQLITE_OK == rc)) {
+    const auto lock = instance->attachLock();
+
+    rc = sqlite3_prepare_v2(
+        instance->db(), sql, -1, &prepared_statement, &leftover_sql);
+    if (rc != SQLITE_OK) {
+      Status s = Status::failure(sqlite3_errmsg(instance->db()));
+      if (prepared_statement != nullptr) {
```

Comment:
No need to check for `nullptr` since `sqlite3_finalize()` can be called with `nullptr`.